Pilot/user_data

```diff
@@ -193,35 +194,56 @@ cp /root/hostkey.pem /scratch/plt/etc/grid-security/hostkey.pem
+# Add plt0102 etc accounts for the payloads that plt can sudo to
+
+# At most one jobagent per logical processor
+processors=`grep '^processor[[:space:]]' /proc/cpuinfo | wc --lines`
+for ((m=0; m < processors; m++))
 do
- for m in 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15
- do
-    mn=`printf '%02dp%02d' $m $n`
-    # add plt0102 etc accounts for the payloads that plt can sudo to
-    /usr/sbin/useradd -m -b /scratch plt$mn
-    /usr/sbin/usermod -a -G plt$mn plt
-    /usr/sbin/usermod -a -G plt plt$mn
-    chmod 0775 /scratch/plt$mn
-    echo "Defaults>plt$mn !requiretty"           >>/etc/sudoers
-    echo "Defaults>plt$mn visiblepw"             >>/etc/sudoers
-    echo "Defaults>plt$mn !env_reset"            >>/etc/sudoers
-    echo "plt ALL = (plt$mn) NOPASSWD: ALL"      >>/etc/sudoers
- done
- n=`expr $n + 1`
+  # Up to 100 successive payloads per jobagent
+  for ((n=0; n < 100; n++))
+  do
+    payloaduser=`printf 'plt%02dp%02d' $m $n`
+    payloaduserid=`printf '1%02d%02d' $m $n`
+
+    # Payload user home directory and dot files
+    mkdir /scratch/$payloaduser
+    cp -f /etc/skel/.*shrc /scratch/$payloaduser
+    cp -f /etc/skel/.bash* /scratch/$payloaduser
+
+    # Add to /etc/passwd and /etc/group
+    echo "$payloaduser:x:$payloaduserid:$payloaduserid::/scratch/$payloaduser:/bin/bash" >>/etc/passwd
+    echo "$payloaduser:x:$payloaduserid:plt" >>/etc/group
+
+    # Add the plt group as a secondary group
+    if [ "$payloaduser" = "plt00p00" ] ; then
+      sed -i "s/^plt:.*/&$payloaduser/" /etc/group
+    else
+      sed -i "s/^plt:.*/&,$payloaduser/" /etc/group
+    fi
+
+    # Ownership and permissions of payload home directory
+    chown -R $payloaduser.$payloaduser /scratch/$payloaduser
+    chmod 0775 /scratch/$payloaduser
+
+    # plt user can sudo to any payload user
+    echo "Defaults>$payloaduser !requiretty"           >>/etc/sudoers
+    echo "Defaults>$payloaduser visiblepw"             >>/etc/sudoers
+    echo "Defaults>$payloaduser !env_reset"            >>/etc/sudoers
+    echo "plt ALL = ($payloaduser) NOPASSWD: ALL"      >>/etc/sudoers
+  done
 done
-) 2>&1 >/var/spool/joboutputs/useradd.log &
 
 cd /scratch/plt
 # Fetch the DIRAC pilot scripts
-if [ '##user_data_option_dirac_pilot_url##' == '' ] ; then
-  wget --no-directories --recursive --no-parent --execute robots=off --reject 'index.html*' https://lhcbproject.web.cern.ch/lhcbproject/Operations/diracpilot/
+if [ '##user_data_option_dirac_pilot_url##' != '' ] ; then
+  wget --no-directories --recursive --no-parent --execute robots=off --reject 'index.html*' --ca-directory=/etc/grid-security/certificates '##user_data_option_dirac_pilot_url##'
+elif [ '##user_data_url##' != '' ] ; then
+  # Remove user_data file name back to final slash
+  user_data_dir=`echo '##user_data_url##' | sed 's:[^/]*$::'`
+  wget --no-directories --recursive --no-parent --execute robots=off --reject 'index.html*' --ca-directory=/etc/grid-security/certificates "$user_data_dir"
```

Comment:
I'll set $X509_CERT_DIR earlier in the user_data shell script but it's all hard coded in effect because the user_data itself is setting everything up. I agree within the pilot itself we should (and do) use $X509_CERT_DIR